### ISSUE TITLE
rustdoc unexpectedly formatted leading numeral as a list

### ISSUE THREAD

**Issue by davepacheco** [rust-lang/rust#92005]
I have a feeling I'm doing something wrong here so I apologize in advance if this is just noise.  But the behavior I'm seeing seems surprising.

I tried this code:

```rust
//! 1. One
//!    1. Two
//!
//! This record means: user 123 has the "admin" role on the "project" with id
//! 234.  (Note.)
```

and I built this with `cargo doc`.

I expected to see this sentence together in one paragraph: "This record means: user 123 has the "admin" role on the "project" with id 234.  (Note.)"

Instead, this happened:

![rustdoc](https://user-images.githubusercontent.com/532688/146440469-9c7b7e62-8e59-4ec6-928d-2d345c09ddb7.png)

Looking at the HTML, it's interpreted the leading "234." as a numbered list.  Thinking this was some Markdown error, I [tried it in the dingus but it does what I expected](https://spec.commonmark.org/dingus/?text=1.%20One%0A%20%20%201.%20Two%0A%0AThis%20record%20means%3A%20user%20123%20has%20the%20%22admin%22%20role%20on%20the%20%22project%22%20with%20id%0A234.%20%20(Note.)).

### Meta

```
$ cargo version
cargo 1.56.0 (4ed5d137b 2021-10-04)
```

```
$ rustdoc --version --verbose
rustdoc 1.56.1 (59eed8a2a 2021-11-01)
binary: rustdoc
commit-hash: 59eed8a2aac0230a8b53e89d4e99d55912ba6b35
commit-date: 2021-11-01
host: x86_64-apple-darwin
release: 1.56.1
LLVM version: 13.0.0
```

On `cargo 1.59.0-nightly (a359ce160 2021-12-14)`, I see almost the same behavior, but the "2" in 234 gets eaten:

![rustdoc-nightly](https://user-images.githubusercontent.com/532688/146441233-835fd323-f632-486f-9e49-5f90ba0b4d7c.png)

**Comment by camelid** [comment]
I imagine this is probably an upstream bug with pulldown-cmark. cc @marcusklaas

**Comment by camelid** [comment]
I wonder if the leading 2 is actually in the HTML but is being cut off by some CSS rules. Can you post the relevant snippet of HTML for both rustdoc versions?

**Comment by davepacheco** [comment]
With `cargo 1.56.0 (4ed5d137b 2021-10-04)`:

```html
<div class="docblock"><ol>
<li>One
<ol>
<li>Two</li>
</ol>
</li>
</ol>
<p>This record means: user 123 has the “admin” role on the “project” with id</p>
<ol start="234">
<li>(Note.)</li>
</ol>
</div>
```

With `cargo 1.59.0-nightly (a359ce160 2021-12-14)`:

```html
<div class="docblock"><ol>
<li>One
<ol>
<li>Two</li>
</ol>
</li>
</ol>
<p>This record means: user 123 has the “admin” role on the “project” with id</p>
<ol start="234">
<li>(Note.)</li>
</ol>
</div>
```

In case it's useful, this is very easily reproduced -- e.g., with https://github.com/davepacheco/rustdoc-test.

**Comment by camelid** [comment]
Thanks! So, indeed, it seems the `2` is being hidden by the CSS; it is actually emitted in the HTML.

**Comment by camelid** [comment]
Assigning priority as [discussed](https://rust-lang.zulipchat.com/#narrow/stream/245100-t-compiler.2Fwg-prioritization.2Falerts/topic/.2392005.20rustdoc.20unexpectedly.20formatted.20leading.20numeral.20as.20a.20.E2.80.A6) in the prioritization working group.

@rustbot label: +P-medium -I-prioritize

**Comment by marcusklaas** [comment]
Ooh, this is an interesting one. While this is certainly surprising, I'm not 100% confident this is a bug in pulldown. I played around in [dingus], the CommonMark reference implementation, and found that it also produced a list for this snippet:
```md
a
1.  d
```

However, if the final line is changed to `2. d` instead, it does not produce a list.

I will have to check the spec, because I can't remember there being different rules for lists that do not start at one.

[dingus]: https://spec.commonmark.org/dingus/

**Comment by marcusklaas** [comment]
It turns out there _is_ a special rule for starting lists with the number one. From the spec:
> In order to solve of unwanted lists in paragraphs with hard-wrapped numerals, we allow only lists starting with 1 to interrupt paragraphs.
And then there follow two tests that test exactly this distinction. Pulldown passes those tests, so I'm confused why it also produces a list for
```md
a
1. b
```
I will create an issue on the pulldown repo and investigate further...

**Comment by marcusklaas** [comment]
We have found the underlying issue and created a PR to address it [here](https://github.com/raphlinus/pulldown-cmark/issues/553). Thanks to @camelid for flagging it with us!

**Comment by camelid** [comment]
Thanks for fixing it!

**Comment by camelid** [comment]
@marcusklaas what is the timeline for the next pulldown release?

**Comment by marcusklaas** [comment]
Very soon!

**Comment by marcusklaas** [comment]
Pulldown version 0.9.0 has just been [released](https://crates.io/crates/pulldown-cmark/0.9.0)!

**Comment by camelid** [comment]
Thanks!

**Comment by camelid** [comment]
This will be fixed once #92252 has landed, which will happen shortly. Closing this issue.
